imgui-sdl2-support/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+    mouse::{Cursor, MouseButton, MouseState, SystemCursor},
+    video::Window,
+    EventPump,
+};
+
+/// State of a single mouse button. Used so that we can detect cases where mouse
+/// press and release occur on the same frame (seems surprisingly frequent on
+/// macOS now...)
+#[derive(Debug, Clone, Default)]
+struct Button {
+    pressed_this_frame: Cell<bool>,
+    state: Cell<bool>,
+}
+
+impl Button {
+    // we can use this in an array initializer, unlike `Default::default()` or a
+    // `const fn new()`.
+    #[allow(clippy::declare_interior_mutable_const)]
+    const INIT: Button = Self {
+        pressed_this_frame: Cell::new(false),
+        state: Cell::new(false),
+    };
+
+    fn set(&self, pressed: bool) {
+        self.state.set(pressed);
+        if pressed {
+            self.pressed_this_frame.set(true);
+        }
+    }
+
+    fn get(&self) -> bool {
+        // If we got a press this frame, record it even if we got a release
+        // too — this way we don't drop mouse clicks where the release comes
+        // in on the same frame as the press. (This mirrors what Dear ImGUI
+        // seems to do in the `imgui_impl_*`)
+        self.pressed_this_frame.replace(false) || self.state.get()
+    }
+}
+
+/// Map an imgui::MouseCursor to an equivalent sdl2::mouse::SystemCursor.
+fn to_sdl_cursor(cursor: MouseCursor) -> SystemCursor {
+    match cursor {
+        MouseCursor::Arrow => SystemCursor::Arrow,
+        MouseCursor::TextInput => SystemCursor::IBeam,
+        MouseCursor::ResizeAll => SystemCursor::SizeAll,
+        MouseCursor::ResizeNS => SystemCursor::SizeNS,
+        MouseCursor::ResizeEW => SystemCursor::SizeWE,
+        MouseCursor::ResizeNESW => SystemCursor::SizeNESW,
+        MouseCursor::ResizeNWSE => SystemCursor::SizeNWSE,
+        MouseCursor::Hand => SystemCursor::Hand,
+        MouseCursor::NotAllowed => SystemCursor::No,
+    }
+}
+
+/// SDL 2 backend platform state.
+pub struct SdlPlatform {
+    cursor_instance: Option<Cursor>, /* to avoid dropping cursor instances */
+    mouse_buttons: [Button; 5],
+}
+
+impl SdlPlatform {
+    fn handle_key_modifier(&mut self, io: &mut Io, keymod: &Mod) {
+        io.key_shift = keymod.intersects(Mod::LSHIFTMOD | Mod::RSHIFTMOD);
+        io.key_ctrl = keymod.intersects(Mod::LCTRLMOD | Mod::RCTRLMOD);
+        io.key_alt = keymod.intersects(Mod::LALTMOD | Mod::RALTMOD);
+        io.key_super = keymod.intersects(Mod::LGUIMOD | Mod::RGUIMOD);
+    }
+
+    fn handle_mouse_button(&mut self, button: &MouseButton, pressed: bool) {
+        match button {
+            MouseButton::Left => self.mouse_buttons[0].set(pressed),
+            MouseButton::Right => self.mouse_buttons[1].set(pressed),
+            MouseButton::Middle => self.mouse_buttons[2].set(pressed),
+            MouseButton::X1 => self.mouse_buttons[3].set(pressed),
+            MouseButton::X2 => self.mouse_buttons[4].set(pressed),
+
+            _ => {}
+        }
+    }
+}
+
+impl SdlPlatform {
```

Comment:
Yes, I usually separate `impl`s into "public" and "private" ones. About the concern, I feel that is a valid point. I think I will just move the private impl below the public one, or should I just merge them?

Reply:
Yeah let's move the private before the public one, and otherwise it's fine. I like developers to leave their own idiosyncrasies on their code, as long as they're not too strange. It's your work, after all